### PR TITLE
post/kafka testing

### PR DESCRIPTION
**Pull request by zhenik**

### CODE REVIEW

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+Examples are built using java and docker. For detailed information, check this repository on [github](https://github.com/sysco-middleware/kafka-testing).
+
+## Agenda
+1. Intro
+2. Unit tests of Kafka Streams application with [kafka-streams-test-utils](https://kafka.apache.org/20/documentation/streams/developer-guide/testing.html)
+3. Integration tests with [EmbeddedKafkaCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils)
+4. End-to-end tests with [Testcontainers](https://www.testcontainers.org)
+
+### 1. Intro
+Many have faced a challenge to test applications that communicate with Kafka cluster. The common issues are how to write unit test, how to automate test for Kafka application in isolation and how to handle end-to-end and system test for whole data pipeline.
```
Comment (timurgen) [inline _posts/2019-06-24-testing-kafka-based-applications.md]: probably plural form "tests" would fit better here

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+1. [Getting Your Feet Wet with Stream Processing - Part 2: Testing Your Streaming Application](https://www.confluent.io/blog/stream-processing-part-2-testing-your-streaming-application)
+2. [Testing Event-Driven Systems](https://www.confluent.io/blog/testing-event-driven-systems)
+
+Main purpose of this article is to share experience of testing Apache Kafka based applications and data pipelines, using two different approaches.
+
+Keywords in use:
+* **Kafka based application** - by Kafka based application I understand any application that uses Kafka API and communicates with Kafka cluster.
+* **Data pipeline** - is a set of Kafka based applications that are connected into a single context.
+
+Examples are built using java and docker. For detailed information, check this repository on [github](https://github.com/sysco-middleware/kafka-testing).
```
Comment (timurgen) [inline _posts/2019-06-24-testing-kafka-based-applications.md:24]: Why (K)afka but (j)ava and (d)ocker?

**Review by timurgen — APPROVED** [review]

File: _posts/2019-06-24-testing-kafka-based-applications.md (hunk 1 above)
Comment (zhenik) [reply _posts/2019-06-24-testing-kafka-based-applications.md]: done

File: _posts/2019-06-24-testing-kafka-based-applications.md (hunk 2 above)
Comment (zhenik) [reply _posts/2019-06-24-testing-kafka-based-applications.md:24]: done

Comment (jeqo) [reply _posts/2019-06-24-testing-kafka-based-applications.md:24]: Is this resolved? I'm still seen java and docker with lowercase, as well as github. 

```suggestion
Examples are built using Java and Docker. For detailed information, check this repository on [GitHub](https://github.com/sysco-middleware/kafka-testing).
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+
+Keywords in use:
+* **Kafka based application** - by kafka based application I understand any application that uses `Kafka API` and communicates with kafka cluster.
+* **Data pipeline** - is a set of Kafka based applications that are connected into a single context.
+
+Examples are built using java and docker. For detailed information, check this repository on [github](https://github.com/sysco-middleware/kafka-testing).
+
+## Agenda
+1. Intro
+2. Unit tests of Kafka Streams application with [kafka-streams-test-utils](https://kafka.apache.org/20/documentation/streams/developer-guide/testing.html)
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]: nit:
```suggestion
2. Unit tests of Kafka Streams application with [kafka-streams-test-utils](https://kafka.apache.org/documentation/streams/developer-guide/testing.html)
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+3. Integration tests with [EmbeddedKafkaCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils)
+4. End-to-end tests with [Testcontainers](https://www.testcontainers.org)
+
+### 1. Intro
+Many have faced a challenge to test applications that communicate with Kafka cluster. The common issues are how to write unit tests, how to automate tests for Kafka application in isolation and how to handle end-to-end and system tests for whole data pipeline.
+
+When I write this article, there are several approaches, some of them not resolved yet:
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]: Rephrase this to "When I wrote this article", or "When this article was written, there were..."

Comment (jeqo) [reply _posts/2019-06-24-testing-kafka-based-applications.md]: nit: Also, "some of them still unresolved"

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+1. [KIP-139: Kafka TestKit library](https://cwiki.apache.org/confluence/display/KAFKA/KIP-139%3A+Kafka+TestKit+library) is still open. Once it will be ready, kafka dev community will have a proper test-kit.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
1. [KIP-139: Kafka TestKit library](https://cwiki.apache.org/confluence/display/KAFKA/KIP-139%3A+Kafka+TestKit+library) is still open. Once it will be ready, the Kafka community will have a proper test-kit.
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+2. [#26 Encapsulate EmbeddedSingleNodeKafkaCluster in a seperately-available maven/gradle/sbt](https://github.com/confluentinc/kafka-streams-examples/issues/26) dep. Kafka cluster in memory. When this post is published, there is an open issue for having release of embedded kafka cluster. [Yeva Byzek](https://www.confluent.io/blog/author/yeva-byzek/) offers
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
2. [#26 Encapsulate EmbeddedSingleNodeKafkaCluster in a seperately-available maven/gradle/sbt](https://github.com/confluentinc/kafka-streams-examples/issues/26) dep. Kafka cluster in memory. When this post was published, there is an open issue for having release of embedded Kafka cluster. [Yeva Byzek](https://www.confluent.io/blog/author/yeva-byzek/) offers 
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+When I write this article, there are several approaches, some of them not resolved yet:
+1. [KIP-139: Kafka TestKit library](https://cwiki.apache.org/confluence/display/KAFKA/KIP-139%3A+Kafka+TestKit+library) is still open. Once it will be ready, kafka dev community will have a proper test-kit.
+2. [#26 Encapsulate EmbeddedSingleNodeKafkaCluster in a seperately-available maven/gradle/sbt](https://github.com/confluentinc/kafka-streams-examples/issues/26) dep. Kafka cluster in memory. When this post is published, there is an open issue for having release of embedded kafka cluster. [Yeva Byzek](https://www.confluent.io/blog/author/yeva-byzek/) offers 
+to use [KafkaEmbeddedCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils) in this [blog-post](https://www.confluent.io/blog/stream-processing-part-2-testing-your-streaming-application). There is no existing release for these utils. I decided to duplicate this part of code base and to build an [internal module](https://github.com/sysco-middleware/kafka-testing/tree/master/embedded-cluster) as an example.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
to use [KafkaEmbeddedCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils) in this [blog-post](https://www.confluent.io/blog/stream-processing-part-2-testing-your-streaming-application). There is no existing release for these libraries. I decided to duplicate this part of code base and to build an [internal module](https://github.com/sysco-middleware/kafka-testing/tree/master/embedded-cluster) as an example.
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+3. [KIP-247 Add public test utils for Kafka Streams](https://cwiki.apache.org/confluence/display/KAFKA/KIP-247%3A+Add+public+test+utils+for+Kafka+Streams). This KIP is ready and has release for unit testing kafka streams topologies - [kafka-streams-test-utils](https://kafka.apache.org/20/documentation/streams/developer-guide/testing.html).
+4. [Testcontainers](https://www.testcontainers.org) - testing with docker containers.
+
+### 2. Unit tests of Kafka Streams application with kafka-streams-test-utils
+[Kafka-streams-test-utils](https://kafka.apache.org/21/documentation/streams/developer-guide/testing.html) is a test-kit for testing stream topologies in memory without need to run Kafka cluster. Test-kit supports stateful operations and avro serialisation/deserialisation via mocking of schema registry.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]: nit:
```suggestion
[Kafka-streams-test-utils](https://kafka.apache.org/21/documentation/streams/developer-guide/testing.html) is a test-kit for testing stream topologies in memory without needing a Kafka cluster. Test-kit supports stateful operations and Avro serialization/deserialization via mocking of [Schema Registry](https://docs.confluent.io/current/schema-registry/index.html).
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+All that you need is TopologyTestDriver, which you can init for each test with topology under test.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
All that is required is to have a `TopologyTestDriver`, which can be initialize on each test case where a topology is under test.
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+
+@After
+public void tearDown() {
+    Optional.ofNullable(testDriver).ifPresent(TopologyTestDriver::close);
+    testDriver = null;
+    properties = null;
+}
+```
+
+I choose for testing topology with stateful operations and avro.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
 I choose for testing topology with stateful operations and Avro. 
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+    final Long amountOfRecordWithSameName = keyValueStore.get("nikita");
+    
+    /** Assert */
+    assertEquals(Long.valueOf(2), amountOfRecordWithSameName);
+}
+```
+
+You can find full test example [here](https://github.com/sysco-middleware/kafka-testing/blob/master/streams-client/src/test/java/no/sysco/testing/kafka/streams/topology/StreamProcessingAvroTest.java) and more examples in [Apache Kafka documentation](https://kafka.apache.org/21/documentation/streams/developer-guide/testing.html). 
+
+Pay attention to issue [#877 - Passing Schema Registry URL twice to instantiate KafkaAvroSerializer](https://github.com/confluentinc/schema-registry/issues/877).
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md:146]: Might be interesting to add a bit of context about this suggestion, like why to pay attention?

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+
+### 3. Integration tests with [EmbeddedKafkaCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils)
+Integration test is a way of how to test services in isolation but with required dependencies. 
+Embedded Kafka cluster combines broker, zookeeper and schema-registry in one. 
+To test asynchronous system I choose [awaitility library](http://www.awaitility.org).
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
To test asynchronous APIs I've chosen [awaitility library](http://www.awaitility.org).
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+  }
+}
+```
+
+### 4. End-to-end tests with [Testcontainers](https://www.testcontainers.org)
+In context of Kafka based applications, end-to-end testing will be applied to  data pipelines to ensure that, first, the data integrity is maintained between applications and, second, data pipelines behave as expected. 
+Testing requires all integrated components of application to be up and running - in order to be able to test them with different scenarios.
+![data pipeline](/images/2019-06-24-testing-kafka-based-applications/data_pipeline.png)
+
+Requirements:
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md:212]: Upper case for software names below

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+* zookeeper 
+* kafka broker 
+* confluent schema registry 
+* http-producer: REST producer 
+* http-materializer: streams application, that downstreams topic data and RPC to another REST service
+* db-mock: REST service and db in memory (use .json file as persistent storage), all-in-one
+
+I describe two approaches of how to prepare container environment for testing. Both of them require a certain start-up sequence, namely: start kafka cluster first, wait until it is ready to serve data, create required topics, register schemas and then start application services.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
I describe two approaches on how to prepare a container environment for testing. Both of them require a certain start-up sequence, namely: start a Kafka cluster first, wait until it is ready to receive requests, create required topics, register schemas and then run application services.
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+}
+```
+
+
+To make containers interact with each other, you should put them with the same network. 
+If one container depends on another container, you might need network alias to setup communication between them. 
+You can provide your own network alias for container or use an already existing one. 
+Aliases can be injected as environmental variable.
+
+Container ports can be mapped to localhost while testing, if container's port is exposed.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
Container ports can be mapped to `localhost` while testing, if container's port is exposed. 
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+    } catch (Exception e) {
+      e.printStackTrace();
+      fail();
+    }
+}
+```
+
+#### Approach 2: Define containers in docker-compose file
+Another approach is to prepare containers environment for testing is to define all the environment information in a docker-compose.yml. In this case Testcontainers manage composition of containers. All services share common network. Name of service uses as a network alias. This approach also supports waiting strategies. 
+Example of how to create topic in a docker-compose file can be found in [confluent examples](https://github.com/confluentinc/examples/blob/f854ac008952346ceaba0d1f1a352788f0572c74/microservices-orders/docker-compose.yml#L182-L215).
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
Example of how to create topic in a `docker-compose` file can be found in [Confluent examples](https://github.com/confluentinc/examples/blob/f854ac008952346ceaba0d1f1a352788f0572c74/microservices-orders/docker-compose.yml#L182-L215).
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+                    .extract()
+                    .as(MessageJsonRepresentation.class);
+
+            assertNotNull(jsonPresentation);
+        });
+}
+```
+
+### Conclusion
+Hopefully, the described approaches and code examples will help make it easier to test kafka clients application. If you would like to contribute, just open a ticket on Github for ideas.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
Hopefully, the described approaches and code examples will help to make testing Kafka client application easier. If you would like to contribute, just open a ticket on GitHub for ideas. 
```

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+                    .as(MessageJsonRepresentation.class);
+
+            assertNotNull(jsonPresentation);
+        });
+}
+```
+
+### Conclusion
+Hopefully, the described approaches and code examples will help make it easier to test kafka clients application. If you would like to contribute, just open a ticket on Github for ideas. 
+Additionally, I would recommend to check Confluent blog that can be helpful in learning more about Kafka.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
Additionally, I would recommend to check [Confluent's blog](https://www.confluent.io/blog/) that can be helpful to learn more about Apache Kafka.
```

File: _posts/2019-06-24-testing-kafka-based-applications.md (hunk 11 above)
Comment (zhenik) [reply _posts/2019-06-24-testing-kafka-based-applications.md:146]: I think context is in issue title. For more info, here is link. I would avoid to put context because it is not related to blog post topic

**Comment by zhenik** [comment]
done.

File: _posts/2019-06-24-testing-kafka-based-applications.md [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+            events.add(record.value());
+  
+          assertEquals(1, events.size());
+      });
+  }
+}
+```
+
+### 4. End-to-end tests with [Testcontainers](https://www.testcontainers.org)
+In context of Kafka based applications, end-to-end testing will be applied to  data pipelines to ensure that, first, the data integrity is maintained between applications and, second, data pipelines behave as expected.
```
Comment (jeqo) [inline _posts/2019-06-24-testing-kafka-based-applications.md]:
```suggestion
In context of Kafka-based applications, end-to-end testing will be applied to data pipelines to ensure that, first, the data integrity is maintained between applications and, second, data pipelines behave as expected.
```
nit: replace others `%s/Kafka based/Kafka-based/g`

**Review by jeqo — APPROVED** [review]
Thanks, @zhenik 
Great job!
